xonsh/tools.py

```diff
@@ -805,6 +805,21 @@ def swap(namespace, name, value, default=NotImplemented):
     else:
         setattr(namespace, name, old)
 
+
+@contextlib.contextmanager
+def swap_values(d, updates, default=NotImplemented):
+    """Updates a dictionary (or other mapping) with values from another mapping,
+    and then restores the original mapping when the context is exited.
+    """
+    old = {k: d.get(k, NotImplemented) for k in updates}
```

Comment:
That is a good idea.  Where did you see that suggested, though? Just out of curiosity...